File: src/main/java/eu/ibagroup/formainframe/utils/crudable/Crudable.java
package eu.ibagroup.formainframe.utils.crudable;

import eu.ibagroup.formainframe.utils.crudable.annotations.Column;
import kotlin.Pair;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Field;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static eu.ibagroup.formainframe.utils.crudable.Utils.*;

public interface Crudable {

  @NotNull <E> Optional<E> add(@NotNull Class<? extends E> rowClass, @NotNull E row);

  @SuppressWarnings("unchecked")
  default <E> Optional<E> add(@NotNull E row) {
    return (Optional<E>) add(row.getClass(), row);
  }

  @NotNull <E> Stream<E> getAll(@NotNull Class<? extends E> rowClass);

  @NotNull <E> Optional<E> update(@NotNull Class<? extends E> rowClass, @NotNull E row);

  @SuppressWarnings("unchecked")
  default <E> Optional<E> update(@NotNull E row) {
    return (Optional<E>) update(row.getClass(), row);
  }

  default @NotNull <E> Optional<E> addOrUpdate(@NotNull Class<? extends E> rowClass, E row) {
    Optional<E> optional = add(rowClass, row);
    if (!optional.isPresent()) {
      optional = update(rowClass, row);
    }
    return optional;
  }

  @SuppressWarnings("unchecked")
  default @NotNull <E> Optional<E> addOrUpdate(@NotNull E row) {
    return (Optional<E>) addOrUpdate(row.getClass(), row);
  }

  @NotNull <E> Optional<E> delete(@NotNull Class<? extends E> rowClass, @NotNull E row);

  @SuppressWarnings("unchecked")
  default @NotNull <E> Optional<E> delete(@NotNull E row) {
    return (Optional<E>) delete(row.getClass(), row);
  }

  @NotNull <E, V> V nextUniqueValue(@NotNull Class<? extends E> rowClass);

  default <E> void replaceGracefully(@NotNull Class<? extends E> rowClass, @NotNull Stream<? extends E> rows) {
    final List<E> current = this.getAll(rowClass).collect(Collectors.toList());
    final List<E> newRows = rows.collect(Collectors.toList());
    applyMergedCollections(rowClass, mergeCollections(current, newRows));
  }

  default <E> void applyMergedCollections(@NotNull Class<? extends E> rowClass, @NotNull MergedCollections<? extends E> mergedCollections) {
    mergedCollections.getToDelete().forEach(e -> delete(rowClass, e));
    mergedCollections.getToUpdate().forEach(e -> update(rowClass, e));
    mergedCollections.getToAdd().forEach(e -> add(rowClass, e));
  }

  static <E> @NotNull MergedCollections<E> mergeCollections(@NotNull Collection<? extends E> oldCollection,
                                                      @NotNull Collection<? extends E> newCollection) {
    final List<Pair<Optional<?>, E>> oldCollectionWithUniqueValues = mapWithUniqueValues(oldCollection);
    final List<Pair<Optional<?>, E>> newCollectionWithUniqueValues = mapWithUniqueValues(newCollection);

    final List<E> toDelete = oldCollectionWithUniqueValues.stream()
        .filter(optionalEPair -> newCollectionWithUniqueValues.stream()
            .noneMatch(optionalEPairNew -> {
              final Optional<?> oldOptional = optionalEPair.getFirst();
              final Optional<?> newOptional = optionalEPairNew.getFirst();
              if (oldOptional.isPresent() && newOptional.isPresent()) {
                return oldOptional.get().equals(newOptional.get());
              }
              return false;
            })
        ).map(Pair::getSecond)
        .collect(Collectors.toList());

    final List<E> toAdd = newCollectionWithUniqueValues.stream()
        .filter(optionalEPair -> oldCollectionWithUniqueValues.stream()
            .noneMatch(optionalEPairOld -> {
              final Optional<?> newOptional = optionalEPair.getFirst();
              final Optional<?> oldOptional = optionalEPairOld.getFirst();
              if (oldOptional.isPresent() && newOptional.isPresent()) {
                return oldOptional.get().equals(newOptional.get());
              }
              return false;
            })
        ).map(Pair::getSecond)
        .collect(Collectors.toList());

    final List<E> toUpdate = newCollectionWithUniqueValues.stream()
        .filter(optionalEPair -> oldCollectionWithUniqueValues.stream()
            .anyMatch(optionalEPairOld -> {
              final Optional<?> newOptional = optionalEPair.getFirst();
              final Optional<?> oldOptional = optionalEPairOld.getFirst();
              if (oldOptional.isPresent() && newOptional.isPresent()) {
                return oldOptional.get().equals(newOptional.get()) && !optionalEPair.getSecond().equals(optionalEPairOld.getSecond());
              }
              return false;
            })
        ).map(Pair::getSecond)
        .collect(Collectors.toList());

    return new MergedCollections<>(
        toAdd, toUpdate, toDelete
    );
  }

  default <E> void addAll(@NotNull Stream<? extends E> rows) {
    rows.forEach(this::add);
  }

  default <E> void addAll(@NotNull Class<? extends E> rowClass,
                          @NotNull Stream<? extends E> rows) {
    rows.forEach(r -> add(rowClass, r));
  }

  @SuppressWarnings("unchecked")
  default @NotNull <E> Stream<E> find(@NotNull Class<? extends E> rowClass,
                                      @NotNull Predicate<? super E> predicate) {
    return (Stream<E>) getAll(rowClass).filter(predicate);
  }

  @SuppressWarnings("unchecked")
  default @NotNull <E, V> Stream<E> getByColumnValue(@NotNull Class<? extends E> rowClass,
                                                     @NotNull String columnName,
                                                     @NotNull V columnValue) {
    return Arrays.stream(getFieldsDeeply(rowClass))
        .filter(getPredicateByColumnName(columnName))
        .findAny()
        .map(field -> (Stream<E>) find(rowClass, e -> {
          try {
            field.setAccessible(true);
            return field.get(e).equals(columnValue);
          } catch (IllegalAccessException illegalAccessException) {
            return false;
          }
        })).orElse(Stream.empty());
  }

  default @NotNull <E, V> Stream<E> getByColumnValue(@NotNull Class<? extends E> rowClass,
                                                     @NotNull Function<? super E, ? extends V> columnGetter,
                                                     @NotNull V columnValue) {
    return find(rowClass, obj -> columnGetter.apply(obj).equals(columnValue));
  }

  default @NotNull <E, V> Stream<E> getByColumnValueFromRow(@NotNull Class<? extends E> rowClass,
                                                            @NotNull Function<? super E, ? extends V> columnGetter,
                                                            @NotNull E anotherRow) {
    return find(rowClass, obj -> columnGetter.apply(obj).equals(columnGetter.apply(anotherRow)));
  }

  @SuppressWarnings("unchecked")
  default @NotNull <E, U> Optional<E> getByUniqueKey(@NotNull Class<? extends E> rowClass,
                                                     @NotNull U uniqueKey) {
    final Field uniqueField = Arrays.stream(getFieldsDeeply(rowClass))
        .filter(getUniqueFieldPredicate())
        .findAny()
        .orElse(null);
    if (uniqueField != null) {
      return (Optional<E>) find(rowClass, e -> {
        try {
          uniqueField.setAccessible(true);
          return uniqueField.get(e).equals(uniqueKey);
        } catch (IllegalAccessException illegalAccessException) {
          return false;
        }
      }).findAny();
    } else {
      return Optional.empty();
    }
  }

  default <E, U> @NotNull Optional<? extends E> deleteByUniqueKey(@NotNull Class<? extends E> rowClass,
                                                                  @NotNull U uniqueKey) {
    final Optional<? extends E> optional = getByUniqueKey(rowClass, uniqueKey);
    if (optional.isPresent()) {
      return delete(optional.get());
    } else {
      return Optional.empty();
    }
  }

  default @NotNull <E> Stream<E> deleteIf(@NotNull Class<? extends E> rowClass,
                                          @NotNull Predicate<? super E> predicate) {
    return find(rowClass, predicate)
        .map(this::delete)
        .filter(Optional::isPresent)
        .map(Optional::get);
  }

  default <E, F> @NotNull Optional<F> getByForeignKey(@NotNull E row, @NotNull String columnName, @NotNull Class<? extends F> foreignRowClass) {
    return getByForeignKeyInternal(this, row, columnName, foreignRowClass);
  }

  default <E, F> @NotNull Optional<F> getByForeignKey(@NotNull E row, @NotNull Class<? extends F> foreignRowClass) {
    return getByForeignKeyInternal(this, row, null, foreignRowClass);
  }

  default <E, F> @NotNull Optional<F> getByForeignKeyDeeply(@NotNull E row, @NotNull Class<? extends F> foreignRowClass) {
    return getByForeignKeyDeeply(row.getClass(), row, foreignRowClass);
  }

  default <E, F> @NotNull Optional<F> getByForeignKeyDeeply(@NotNull Class<? extends E> rowClass, @NotNull E row, @NotNull Class<? extends F> foreignRowClass) {
    return Optional.ofNullable(getByForeignKeyDeeplyInternal(this, row, foreignRowClass));
  }

  static @NotNull <E, U> Optional<U> getUniqueValue(@NotNull E row, @NotNull Class<? extends U> uniqueValueClass) {
    return Arrays.stream(getFieldsDeeply(row))
        .filter(field -> {
          field.setAccessible(true);
          return field.isAnnotationPresent(Column.class) && field.getAnnotation(Column.class).unique();
        })
        .findAny()
        .map(field -> {
          try {
            final Object uniqueObj = field.get(row);
            if (uniqueValueClass.isAssignableFrom(uniqueObj.getClass())) {
              return uniqueValueClass.cast(uniqueObj);
            }
          } catch (IllegalAccessException ignored) {
          }
          return null;
        });
  }

  static <E> @NotNull Optional<?> getUniqueValueForRow(@NotNull E row) {
    return Arrays.stream(getFieldsDeeply(row))
        .filter(field -> {
          field.setAccessible(true);
          return field.isAnnotationPresent(Column.class) && field.getAnnotation(Column.class).unique();
        })
        .findAny()
        .map(field -> {
          field.setAccessible(true);
          try {
            return field.get(row);
          } catch (IllegalAccessException e) {
            return null;
          }
        });
  }

}

File: src/main/kotlin/eu/ibagroup/formainframe/utils/crudable/Crudable.kt
/*
 * Copyright (c) 2020-2024 IBA Group.
 *
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   IBA Group
 *   Zowe Community
 */

package eu.ibagroup.formainframe.utils.crudable

import eu.ibagroup.formainframe.analytics.AnalyticsService
import eu.ibagroup.formainframe.analytics.events.ActionType
import eu.ibagroup.formainframe.analytics.events.ConnectionEvent
import eu.ibagroup.formainframe.config.connect.ConnectionConfig
import eu.ibagroup.formainframe.utils.crudable.Utils.castToColumnOrNull
import eu.ibagroup.formainframe.utils.nullable
import java.lang.reflect.Field
import java.util.*
import java.util.function.Predicate
import java.util.stream.Collectors
import java.util.stream.Stream

inline fun <reified E : Any> Crudable.find(noinline predicate: (E) -> Boolean): Stream<E> {
  return this.find(E::class.java, predicate)
}

inline fun <reified E : Any> Crudable.getAll(): Stream<E> {
  return this.getAll(E::class.java)
}

inline fun <reified E : Any> Crudable.getByColumnLambda(row: E, crossinline lambdaGetter: (E) -> Any): Stream<E> {
  return this.find { lambdaGetter(it) == lambdaGetter(row) }
}

inline fun <reified E : Any> Crudable.getByUniqueKey(uniqueKey: Any): E? {
  return this.getByUniqueKey(E::class.java, uniqueKey).nullable
}

inline fun <reified E : Any, reified V : Any> Crudable.nextUniqueValue(): V {
  return this.nextUniqueValue(E::class.java)
}

interface AddFilter {
  operator fun <T : Any> invoke(clazz: Class<out T>, addingRow: T): Boolean = true
}

interface UpdateFilter {
  operator fun <T : Any> invoke(clazz: Class<out T>, currentRow: T, updatingRow: T): Boolean = true
}

/**
 * Interface that describes the basic interaction with CRUDable object
 */
interface Crudable {
  fun <E : Any> add(rowClass: Class<out E>, row: E): Optional<E>
  fun <E : Any> add(row: E): Optional<E>? {
    return add(row.javaClass, row)
  }

  /**
   * Get all rows of the row class from crudable
   * @param rowClass the row class to get all rows by
   */
  fun <E : Any> getAll(rowClass: Class<out E?>): Stream<E>
  fun <E : Any> update(rowClass: Class<out E>, row: E): Optional<E>
  fun <E : Any> update(row: E): Optional<E>? {
    return update(row.javaClass, row)
  }

  fun <E : Any> addOrUpdate(rowClass: Class<out E>, row: E): Optional<E> {
    var optional = add(rowClass, row)
    if (optional.isEmpty) {
      optional = update(rowClass, row)
    }
    return optional
  }

  fun <E : Any> addOrUpdate(row: E): Optional<E> {
    return addOrUpdate(row.javaClass, row)
  }

  fun <E : Any> delete(rowClass: Class<out E>, row: E): Optional<E>
  fun <E : Any> delete(row: E): Optional<E> {
    return delete(row.javaClass, row)
  }

  fun <E : Any?, V : Any?> nextUniqueValue(rowClass: Class<out E?>): V

  /**
   * Replace the rows of the class with the new provided rows
   *
   * @param rowClass row class to replace rows
   * @param rows     new rows to replace the old ones
   */
  fun <E : Any> replaceGracefully(rowClass: Class<out E?>, rows: Stream<out E?>) {
    val current = this.getAll(rowClass).collect(Collectors.toList())
    val newRows = rows.toList().filterNotNull()
    applyMergedCollections(rowClass, mergeCollections(current, newRows))
  }

  /**
   * Perform actions on the row class with the new rows
   *
   * @param rowClass          row class to perform actions on
   * @param mergedCollections merged row collections to perform actions by
   */
  fun <E : Any> applyMergedCollections(
    rowClass: Class<out E>,
    mergedCollections: MergedCollections<out E>
  ) {
    mergedCollections.toDelete.forEach { e: E -> delete(rowClass, e) }
    mergedCollections.toUpdate.forEach { e: E -> update(rowClass, e) }
    mergedCollections.toAdd.forEach { e: E ->
      add(rowClass, e)
      if (e is ConnectionConfig) {
        AnalyticsService.getService().trackAnalyticsEvent(ConnectionEvent(ActionType.CREATE))
      }
    }
  }

  fun <E : Any> find(
    rowClass: Class<out E?>,
    predicate: Predicate<in E>
  ): Stream<E> {
    return getAll(rowClass).filter(predicate) as Stream<E>
  }

  /**
   * Get field by unique key
   *
   * @param rowClass  row class to get fields by
   * @param uniqueKey unique key to get field by
   * @return the field or nothing
   */
  fun <E : Any, U> getByUniqueKey(
    rowClass: Class<out E?>,
    uniqueKey: U
  ): Optional<E> {
    val uniqueField = Arrays.stream(Utils.getFieldsDeeply(rowClass))
      .filter(Utils.uniqueFieldPredicate)
      .findAny()
      .orElse(null)
    return if (uniqueField != null) {
      find(rowClass) { e: E ->
        try {
          uniqueField.isAccessible = true
          return@find uniqueField[e] == uniqueKey
        } catch (illegalAccessException: IllegalAccessException) {
          return@find false
        }
      }.findAny() as Optional<E>
    } else {
      Optional.empty<E>()
    }
  }

  fun <E : Any, U> deleteByUniqueKey(
    rowClass: Class<out E?>,
    uniqueKey: U
  ): Optional<out E> {
    val optional: Optional<out E> = getByUniqueKey(rowClass, uniqueKey)
    return if (optional.isPresent) {
      delete(optional.get())
    } else {
      Optional.empty()
    }
  }

  fun <E : Any, F : Any> getByForeignKey(row: E, foreignRowClass: Class<out F>): Optional<F>? {
    return Utils.getByForeignKeyInternal(this, row, null, foreignRowClass)
  }

  companion object {
    /**
     * Merge the old and the new row collections, so they will be removed, updated or added respectively
     *
     * @param oldCollection the old rows collection
     * @param newCollection the new rows collection
     */
    fun <E : Any> mergeCollections(
      oldCollection: Collection<E>,
      newCollection: Collection<E>
    ): MergedCollections<E> {
      val oldCollectionWithUniqueValues = Utils.mapWithUniqueValues(oldCollection)
      val newCollectionWithUniqueValues = Utils.mapWithUniqueValues(newCollection)
      val toDelete = oldCollectionWithUniqueValues.stream()
        .filter { (oldOptional): Pair<Optional<*>, E> ->
          newCollectionWithUniqueValues.stream()
            .noneMatch { (newOptional): Pair<Optional<*>, E> ->
              if (oldOptional.isPresent && newOptional.isPresent) {
                return@noneMatch oldOptional.get() == newOptional.get()
              }
              false
            }
        }
        .map { pair: Pair<*, E> -> pair.second }
        .collect(Collectors.toList())
      val toAdd = newCollectionWithUniqueValues.stream()
        .filter { (newOptional): Pair<Optional<*>, E> ->
          oldCollectionWithUniqueValues.stream()
            .noneMatch { (oldOptional): Pair<Optional<*>, E> ->
              if (oldOptional.isPresent && newOptional.isPresent) {
                return@noneMatch oldOptional.get() == newOptional.get()
              }
              false
            }
        }
        .map { pair: Pair<*, E> -> pair.second }
        .collect(Collectors.toList())
      val toUpdate = newCollectionWithUniqueValues.stream()
        .filter { (newOptional, second): Pair<Optional<*>, E> ->
          oldCollectionWithUniqueValues.stream()
            .anyMatch { (oldOptional, second1): Pair<Optional<*>, E> ->
              if (oldOptional.isPresent && newOptional.isPresent) {
                return@anyMatch oldOptional.get() == newOptional.get() && second != second1
              }
              false
            }
        }
        .map { pair: Pair<*, E> -> pair.second }
        .collect(Collectors.toList())
      return MergedCollections(
        toAdd, toUpdate, toDelete
      )
    }

    /**
     * Get column value with column "unique" attribute set to "true"
     * @param row the row to get the unique column value from
     */
    fun <E : Any> getUniqueValueForRow(row: E): Optional<*> {
      return Arrays.stream(Utils.getFieldsDeeply(row.javaClass))
        .filter { field: Field ->
          field.isAccessible = true
          field.castToColumnOrNull()?.unique ?: false
        }
        .findAny()
        .map { field: Field ->
          field.isAccessible = true
          try {
            return@map field[row]
          } catch (e: IllegalAccessException) {
            return@map null
          }
        }
    }
  }
}
